### PR TITLE
(PUP-5730) Add iterative step function

### PR DESCRIPTION
**Pull request by thallgren**
This PR adds the new function `step(iterable, n)`. The function takes two arguments, an iterable and an integer, and an optional block.

If a block is given, it will be called with each nth element of the given argument.

Without a block, the function will return a new Iterable that will yield each nth element of the given argument.

### CODE REVIEW

**Comment by hlindberg** [comment]
only some doc and test things to address - then ready to be merged

File: lib/puppet/functions/step.rb [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# Provides stepping with given interval over elements in an iterable and optionally runs a
+# [lambda](http://docs.puppetlabs.com/puppet/latest/reference/lang_lambdas.html) for each
+# element.
+#
+# This function takes two to three arguments.
+#
+# 1. An iterable that the function will iterate over.
```
Comment (hlindberg) [inline lib/puppet/functions/step.rb]: lower case iterable

File: lib/puppet/functions/step.rb [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+# # For the array $data, return an array, set to the first element and each 5th successor element, in reverse
+# # order multiplied by 10
+# $data = Integer[0,20]
+# $transformed_data = $data.step(5).map |$item| { $item * 10 }
+# # $transformed_data contains [0,50,100,150,200]
+# ~~~
+#
+# @example The same example using `step` function chained with a `map` in alternative syntax
+#
+# # For the array $data, return an array containing the first and each 5th successor, in reverse order, multiplied by 10
```
Comment (hlindberg) [inline lib/puppet/functions/step.rb]: "return" => "produce"

I prefer "produce" unless specifically talking about a function or lambda that returns - here the sentence is general as it says "for the array $data, return".

**Comment by hlindberg** [comment]
only picky doc comments remain before merge

File: lib/puppet/functions/step.rb (hunk 2 above)
Comment (thallgren) [inline lib/puppet/functions/step.rb]: I would argue that "return" is more precise. A function can "produce" other things than a return value (notices for instance). Also, the term "return" happens to coincide with the @return annotation and is a very well known concept. Why confuse things by using "produce"?

**Pull request merged**